File: 2024/03.cpp
#include <fstream>
#include <iostream>
#include <regex>
#include <string>

int main() {
  std::ifstream input_file("input");
  std::string line;
  int total = 0, conditional = 0;
  static const std::regex mul(R"(mul\((\d{1,3}),(\d{1,3})\))"),
      command(R"(don't\(\)|do\(\))");

  bool enabled = true;
  while (std::getline(input_file, line)) {
    for (std::smatch m; std::regex_search(line, m, mul); line = m.suffix()) {
      const int prod = std::stoi(m[1]) * std::stoi(m[2]);
      total += prod;

      const std::string prefix = m.prefix().str();
      if (std::smatch prefix_m; std::regex_search(prefix, prefix_m, command)) {
        enabled = prefix_m[0] == "do()";
      }
      if (enabled) {
        conditional += prod;
      }
    }
  }
  std::cout << total << '\n' << conditional << std::endl;
}

File: 2024/03.rs
use regex::Regex;
use std::error::Error;

const INPUT: &str = include_str!("input");

fn main() -> Result<(), Box<dyn Error>> {
    let mul = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)")?;
    let command = Regex::new(r"(don't\(\)|do\(\))")?;

    let (mut total, mut conditional) = (0, 0);
    let mut enabled = true;
    for line in INPUT.lines() {
        let mut prev_match_end = 0;
        for (range, factors) in mul
            .captures_iter(line)
            .map(|c| (c.get(0).unwrap().range(), c.extract::<2>().1))
        {
            let product: i32 = factors.iter().map(|f| f.parse::<i32>().unwrap()).product();
            total += product;

            if let Some(command_match) = command.captures(&line[prev_match_end..range.start]) {
                enabled = command_match.get(0).ok_or("")?.as_str() == "do()";
            }
            if enabled {
                conditional += product;
            }
            prev_match_end = range.end;
        }
    }

    println!("{total}\n{conditional}");
    Ok(())
}
